Add table showing complexity of methods for Input

ShadowMitia opened the pull request:
# Objective

Fixes #10106 

Adds a table showing computational complexity, as seen for Query (and similar to std::collections docs).

## Solution

Add the complexity table

---

## Changelog

- Add complexity table for Input methods

ShadowMitia commented:
I've added information on all methods, hope this is ok.

I've tried analyzing the source and I've also used the [std::collections doc](https://doc.rust-lang.org/stable/std/collections/index.html) on performance also as a reference.

I'm not completely sure on the values for just_pressed/just_released/pressed, which all have a call to `contain`.
I'm kind of assuming it is comparable to a `get` complexity as seen in the collections docs.

alice-i-cecile commented on crates/bevy_input/src/input.rs:
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
+///
+/// | **[`Input`] operations**          | **Computational complexity**       |
+/// |-----------------------------------|------------------------------------|
+/// | [`Input::any_just_pressed`]       | *O*(m*n)                           |
```
I think n * m will be slightly easier to read.

alice-i-cecile commented on crates/bevy_input/src/input.rs:
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
+/// For all operations, the following conventions are used:
+/// - **n** is the number of stored inputs.
+/// - **m** is the number of elements of another collection involved in the operation.
+/// - **\***-suffix denotes an amortized cost.
```
Given that this isn't used here, I think it should be cut from this table.

ShadowMitia replied:
Sorry I don't understand, I'm using all the symbols?
To be sure, this **\*** is for the * symbol.

alice-i-cecile replied:
My bad, I see now. It was hard to follow in markdown.

alice-i-cecile reviewed (approved):
Nits on the wording, but nothing blocking. All of the listed time complexities appear correct to me.

james7132 commented on crates/bevy_input/src/input.rs:
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
+/// | [`Input::clear_just_pressed`]     | *O*(1)~                            |
+/// | [`Input::clear_just_released`]    | *O*(1)~                            |
+/// | [`Input::reset_all`]              | *O*(n)                             |
+/// | [`Input::clear`]                  | *O*(n)                             |
```
Is this accurate? HashSet::clear is O(1), no?

ShadowMitia replied:
I'll double-check properly when I can, and I could be wrong, but on clear the hashtable goes through all elements to drop them one at a time.
Maybe this is only true for types with a custom drop?

ShadowMitia replied:
HashSets are based on HashMaps, and both use an implementation based on the hasbrown crate.
If you follow down the clear() to where it is actually doing things, you get here:
https://github.com/rust-lang/hashbrown/blob/f2e62124cd947b5e2309dd6a24c7e422932aae97/src/raw/mod.rs#L1091
Which doesn't do much except call drop_elements which is defined here:
https://github.com/rust-lang/hashbrown/blob/f2e62124cd947b5e2309dd6a24c7e422932aae97/src/raw/mod.rs#L2369
That last one I consider to be `O(n)` hence the information put into the table. But its only run for cases where the type needs to be dropped.

So I guess I need to specify both cases? Or should we say it is `O(n)` worse case?

matiqo15 replied:
I would go for `O(n)` because it's standard to refer to the worst case scenario when talking about time complexity.

nelsontkq replied:
Looks like [it's marked as O(1)~ in the std docs](https://doc.rust-lang.org/stable/std/collections/index.html#maps).

I think it's more useful to know it's O(1) in 99.999999% of cases, and most likely in this case since there's probably less than 20 things in each of those sets.

mockersf commented on crates/bevy_input/src/input.rs:
```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
+///
+/// For all operations, the following conventions are used:
+/// - **n** is the number of stored inputs.
+/// - **m** is the number of input types.
```
I'm not sure of the meaning of "input types", there's only one type `T`

nelsontkq replied:
```suggestion
/// - **m** is the number of input parameters passed to the method.
```

ShadowMitia commented:
Updated with all suggestions and rebased on latest changes.
I still think the clear method should be documented as `O(n)`, unless there is a good reason not to :smile: 
Should be good to go otherwise!
